src/core/struct_tree.js

```diff
@@ -67,6 +69,439 @@ class StructTreeRoot {
+
+        if (title) {
+          tagDict.set("T", title);
+        }
+        if (lang) {
+          tagDict.set("Lang", lang);
+        }
+        if (alt) {
+          tagDict.set("Alt", alt);
+        }
+        if (expanded) {
+          tagDict.set("E", expanded);
+        }
+        if (actualText) {
+          tagDict.set("ActualText", actualText);
+        }
+
+        if (structTreeParent) {
+          await this.#updateParentTag(
+            structTreeParent,
+            tagDict,
+            tagRef,
+            xref,
+            newRefs,
+            buffer
+          );
+        } else {
+          tagDict.set("P", structTreeRootRef);
+        }
+
+        const objDict = new Dict(xref);
+        tagDict.set("K", objDict);
+        objDict.set("Type", objr);
+        objDict.set("Pg", pageRef);
+        objDict.set("Obj", ref);
+
+        buffer.length = 0;
+        await writeObject(tagRef, tagDict, buffer, xref);
+        newRefs.push({ ref: tagRef, data: buffer.join("") });
+
+        nums.push(parentTreeId, tagRef);
+        kids.push(tagRef);
+      }
+    }
+    return nextKey + 1;
+  }
+
+  static #collectParents(elements, xref, pageDict, pageIndex, parentTree) {
+    const idToElement = new Map();
+    for (const element of elements) {
+      if (element.structTreeParentId) {
+        const id = parseInt(element.structTreeParentId.split("_mc")[1], 10);
+        idToElement.set(id, element);
+      }
+    }
+
+    const id = pageDict.get("StructParents");
+    const numberTree = new NumberTree(parentTree, xref);
+    const parentArray = numberTree.get(id);
+    if (Array.isArray(parentArray)) {
+      for (const kidRef of parentArray) {
+        if (!(kidRef instanceof Ref)) {
+          continue;
+        }
+        const pageKid = xref.fetch(kidRef);
+        const k = pageKid.get("K");
+        if (Number.isInteger(k)) {
+          const element = idToElement.get(k);
+          if (element) {
+            const parentRef = pageKid.getRaw("P");
+            const parentDict = xref.fetchIfRef(parentRef);
+            if (parentRef instanceof Ref && parentDict instanceof Dict) {
+              // It should always the case, but we check just in case.
+              element.structTreeParent = { ref: kidRef, dict: pageKid };
+            }
+          }
+          continue;
+        }
+
+        if (!Array.isArray(k)) {
+          continue;
+        }
+        for (const kid of k) {
+          if (Number.isInteger(kid)) {
+            const element = idToElement.get(kid);
+            if (element) {
+              const parentRef = pageKid.getRaw("P");
+              const parentDict = xref.fetchIfRef(parentRef);
+              if (parentRef instanceof Ref && parentDict instanceof Dict) {
+                // It should always the case, but we check just in case.
+                element.structTreeParent = { ref: kidRef, dict: pageKid };
+              }
+              break;
+            }
```

Comment:
With the helper function suggested above, this could become:
```suggestion
            if (updateElement(kid)) {
              break;
            }
```